syntax = "proto3";

option go_package = ".;pb";

import "google/protobuf/empty.proto";

message RouteTable {
  string network_name = 1;
  uint32 id = 2;
  string name = 3;
}

message RouteTableIdentifier {
  uint32 id = 1;
}

message ListRouteTablesRequest {
  string network_name = 1;
}

message ListRouteTablesResponse {
  repeated RouteTable route_tables = 1;
}

message CreateRouteTableRequest {
  RouteTable route_table = 1;
}

message RouteIdentifier {
  uint32 route_table_id = 1;
  string destination = 2;
}

message Route {
  uint32 route_table_id = 1;
  string destination = 2;
  repeated string gateways = 3;
}

message ListRoutesRequest {
  uint32 route_table_id = 1;
}

message ListRoutesResponse {
  repeated Route routes = 1;
}

message PutRouteRequest {
  Route route = 1;
}

service RouteService {
  rpc GetRouteTable(RouteTableIdentifier) returns (RouteTable) {}
  rpc ListRouteTables(ListRouteTablesRequest) returns (ListRouteTablesResponse) {}
  rpc CreateRouteTable(CreateRouteTableRequest) returns (RouteTable) {}
  rpc DeleteRouteTable(RouteTableIdentifier) returns (google.protobuf.Empty) {}

  rpc GetRoute(RouteIdentifier) returns (Route) {}
  rpc ListRoutes(ListRoutesRequest) returns (ListRoutesResponse) {}
  rpc PutRoute(PutRouteRequest) returns (Route) {}
  rpc DeleteRoute(RouteIdentifier) returns (google.protobuf.Empty) {}
}
